src/realm/sync/noinst/client_reset.cpp

```diff
@@ -985,38 +991,85 @@ LocalVersionIDs perform_client_reset_diff(DBRef db_local, DBRef db_remote, sync:
+            RecoverLocalChangesetsHandler handler{*wt_remote, *frozen_pre_local_state, logger};
+            handler.process_changesets(local_changes, {}); // throws on error
+            ClientReplication* client_repl = dynamic_cast<ClientReplication*>(wt_remote->get_replication());
+            REALM_ASSERT_RELEASE(client_repl);
+            ChangesetEncoder& encoder = client_repl->get_instruction_encoder();
+            const sync::ChangesetEncoder::Buffer& buffer = encoder.buffer();
+            recovered_changeset = {buffer.data(), buffer.size()};
+        }
+        else {
+            // In FLX recovery, save a copy of the pending subscriptions for later. This
+            // needs to be done before they are wiped out by remake_active_subscription()
+            pending_subscriptions = sub_store->get_pending_subscriptions();
+        }
     }
 
+    // transform the local Realm such that all public tables become identical to the remote Realm
     transfer_group(*wt_remote, *wt_local, logger);
 
+    // now that the state of the fresh and local Realms are identical,
+    // reset the local sync history and steal the fresh Realm's ident
     history_local->set_client_reset_adjustments(current_version_local, client_file_ident, fresh_server_version,
                                                 recovered_changeset);
 
     // Finally, the local Realm is committed. The changes to the remote Realm are discarded.
     wt_remote->rollback_and_continue_as_read();
     wt_local->commit_and_continue_as_read();
 
-    // In DiscardLocal mode, only the active subscription set is preserved.
+    // If in FLX mode, make a copy of the active subscription set and mark it as
+    // complete. This will cause all other subscription sets to become superceded. If
+    // this is DiscardLocal mode, only the active subscription set is preserved, so we
+    // are done. If this is recovery mode, any pending subscriptions will be reapplied
+    // on top below.
     remake_active_subscription();
 
+    if (recover_local_changes && sub_store) {
+        // In FLX recovery, transform the local state into the remote state first, then remake the
+        // active subscription as complete, and apply local changes interleaved with pending subscriptions
+        // in separate commits as needed. This has the consequence that there may be extra notifications
+        // along the way to the final state, but since separate commits are necessary, this is unavoidable.
+        try {
+            wt_local = db_local->start_write();
+            RecoverLocalChangesetsHandler handler{*wt_local, *frozen_pre_local_state, logger};
+            handler.process_changesets(local_changes, std::move(pending_subscriptions)); // throws on error
+            wt_local->commit_and_continue_as_read();
+        }
+        catch (const std::exception& e) {
+            // If there is a fatal error while recovering the changesets in FLX mode,
```

Comment:
it would be nice to test this scenario if possible (did not check the tests yet :))

Reply:
I thought this was going to be nearly impossible to test, but I found a way and added some tests for it.